### PR TITLE
QDEIM with QR via scalapack

### PR DESCRIPTION
**Pull request by dylan-copeland**
QDEIM is implemented, as an alternative to DEIM and GNAT, with oversampling done by GappyPOD+E (see https://arxiv.org/abs/1808.10473). The first samples (numbering up to the basis dimension) for QDEIM are computed using parallel QR with column pivoting for the transposed basis matrix, via `pdgeqpf` in scalapack.

The only validation of this code is its application in Laghos, which has some good results but does not truly validate the QR or QDEIM implementations.

### CODE REVIEW

**Comment by chldkdtn** [comment]
@dylan-copeland This is awesome! Is this ready for review?

**Comment by dylan-copeland** [comment]
> @dylan-copeland This is awesome! Is this ready for review?

@chldkdtn Not ready yet. There will also be a small PR for Laghos to use QDEIM. We should start using PR labels (e.g. "Work-in-progress", "Ready-for-review").

File: Matrix.C
Comment (kevinhkhuynh) [inline Matrix.C:341]: In Ardra, there is an example of a distributed temporal basis matrix being pointwise multiplied with a non-distributed vector, and so setting this to VERIFY breaks the code. What do we think about only setting dimension checks to VERIFY (which is already done in the other merged PR) and leaving distributed checks as ASSERT?

Comment (dylan-copeland) [reply Matrix.C:341]: If the condition is `false`, with either `ASSERT` or `VERIFY`, then something is wrong. It sounds like the check itself is wrong, if it works in a case where the condition is `false`. That is, if a distributed matrix times a non-distributed vector works, then we should just remove the check entirely. In this case, if in Ardra you have a distributed matrix, and one of its rows is multiplied by a non-distributed vector, I'm not sure how that could work. The row index is local, so how is it specified in parallel? This function is supposed to multiply the vector by the row, which is stored entirely on the process owning that row, but that process would be the only one having the correct result and would need to do `MPI_Bcast` to broadcast the result to all processes. I'm guessing there is a peculiar usage that makes this work, but it doesn't sound right in general. Can you provide more details?

Comment (kevinhkhuynh) [reply Matrix.C:341]: @chldkdtn Perhaps Youngsoo can provide more accurate information on the meaning of the Ardra code and whether it can be rewritten to not break the ASSERT/VERIFY. 

Currently, the code is working, and for space-time ROM, there is an non-distributed initialized vector called carom_red_soln_init. It seems this vector can be distributed, but right now, it is not. The full order solution is projected onto the reduced order vector via: projectLiftedSolution(carom_red_soln_init, carom_soln, spatialbasis);. Then the reduced order solution is scaled by the temporal basis via: scaleWithTemporalBasis(carom_red_soln_init, temporalbasis, 0), where only row = 0 is chosen. This scaling function uses pointwise_mult and throws an error due to temporalbasis failing !distributed().

Comment (dylan-copeland) [reply Matrix.C:341]: So local row 0 on every process scales the same undistributed vector on every process, with different results on every process? Is it used on every process?

Comment (kevinhkhuynh) [reply Matrix.C:341]: Hi Dylan. You were right. It seems that temporal basis is distributed in name, but actually not distributed as all the processors have the same temporal basis. However, the spatial basis is distributed. And so, since they were created from one basis generator, both were set to distributed. I don't think libROM currently allows to set rightSV to not distributed while leftSV is.

Comment (chldkdtn) [reply Matrix.C:341]: Looks like both of you have figured out what is going on with temporal basis. It is currently duplicated over the processors. If we can generalize the case, e.g., allowing distributed spatial basis and non-distributed temporal basis, that would be wonderful.

Comment (dylan-copeland) [reply Matrix.C:341]: It sounds like the temporal basis should be non-distributed, in which case no changes to the `Matrix` class would be needed (just changes in the basis generator). I am not yet familiar with this issue or a test that could verify a libROM change is working, so I propose just keeping this as an `ASSERT` in this PR and letting @kevinhkhuynh fix this for Ardra in another PR. Does that sound good @kevinhkhuynh @chldkdtn ?

Comment (chldkdtn) [reply Matrix.C:341]: @dylan-copeland @kevinhkhuynh Please remind me again. ASSERT becomes only active with debug mode, while VERIFY is active both with optimized and debug modes, right?

Comment (kevinhkhuynh) [reply Matrix.C:341]: @dylan-copeland That sounds good to me.

@chldkdtn ASSERT is only active in debug mode, while VERIFY is active in all modes.

File: QDEIM.C
Comment (chldkdtn) [reply QDEIM.C:58]: > I have a similar line in GNAT that is failing due to num_samples possibly being greater than f_basis->numRows for parallel runs since numRows() only returns the number of rows on the current processor. I've created a numDistributedRows function in the tests/gnatTest PR. Perhaps once we merge that PR, we can use that function here too.

@kevinhkhuynh Just to make sure, numDistributedRows return global number of rows, right?

**Review by siuwuncheung — APPROVED** [review]

**Comment by dylan-copeland** [comment]
@chldkdtn @kevinhkhuynh reminder that this is the only QDEIM PR waiting for your review.

File: QDEIM.C
Comment (kevinhkhuynh) [reply QDEIM.C:58]: @dylan-copeland I can make the change to this line in my test_qdeim branch once test_gnat is finished next week, so this PR isn't held up. How does that sound?

**Review by kevinhkhuynh — APPROVED** [review]

File: QDEIM.C
Comment (chldkdtn) [reply QDEIM.C:58]: That sounds good.

**Review by chldkdtn — APPROVED** [review]

**Pull request merged**
